aten/src/ATen/native/sparse/cuda/SparseFactories.cu

```diff
@@ -0,0 +1,233 @@
+// It does not matter if Q values fit  on any diagonal of S, or if S has no
+// O[i]th diagonal (those values/diagonals are simply skipped)
+// --------------------------------------------------------------------
+
+namespace {
+
+template <typename scalar_t>
+C10_LAUNCH_BOUNDS_1(1024)
+__global__ void _spdiags_sparse_coo_cuda_kernel(
+    int64_t total_diags,
+    const TensorInfo<int64_t, int64_t> offsets_ti,
+    const TensorInfo<scalar_t, int64_t> diagonals_ti,
+    const TensorInfo<int64_t, int64_t> nnz_per_diag_ti,
+    const TensorInfo<int64_t, int64_t> nnz_prefix_ti,
+    TensorInfo<int64_t, int64_t> indices_out_ti,
+    TensorInfo<scalar_t, int64_t> values_out_ti) {
+  // First input diagonal to handle
+  const int64_t d_0 = blockIdx.x * blockDim.x + threadIdx.x;
+  int64_t diag_stride0 = diagonals_ti.strides[0];
+  int64_t indices_stride0 = indices_out_ti.strides[0];
+  int64_t grid_stride = blockDim.x * gridDim.x;
+
+  for (int64_t di = d_0; di < total_diags; di += grid_stride) {
+    int64_t n_out = nnz_per_diag_ti.data[di];
+    if (n_out > 0) {
+      int64_t offset = offsets_ti.data[di];
+      int64_t col_begin = ::max(offset, int64_t{0});
+      int64_t row_begin = col_begin - offset;
+      int64_t read_begin = (diag_stride0 * di) + col_begin;
+      int64_t write_begin = nnz_prefix_ti.data[di];
+      for (int64_t i = 0; i < n_out; ++i) {
+        values_out_ti.data[write_begin + i] = diagonals_ti.data[read_begin + i];
+        indices_out_ti.data[write_begin + i] = row_begin + i;
+        indices_out_ti.data[indices_stride0 + write_begin + i] = col_begin + i;
+      }
+    }
+  }
+}
+} // namespace
+
+SparseTensor spdiags_sparse_cuda(
+    const Tensor& diagonals,
+    const Tensor& offsets,
+    IntArrayRef shape,
+    c10::optional<Layout> layout) {
+  TORCH_CHECK(diagonals.dim() == 2, "Diagonals must be 2d");
+  TORCH_CHECK(shape.size() == 2, "Output shape must be 2d");
+  TORCH_CHECK(offsets.dim() == 1, "Offsets must be 1d");
+  TORCH_CHECK(
+      diagonals.size(0) == offsets.size(0),
+      "Number of diagonals (",
+      diagonals.size(0),
+      ") does not match the number of offsets (",
+      offsets.size(0),
+      ")");
+  if (layout) {
+    TORCH_CHECK(
+        (*layout == Layout::Sparse) || (*layout == Layout::SparseCsc) ||
+            (*layout == Layout::SparseCsr),
+        "Only output layouts (",
+        Layout::Sparse,
+        ", ",
+        Layout::SparseCsc,
+        ", and ",
+        Layout::SparseCsr,
+        ") are supported");
+  }
+
+  const int64_t n_row_out = shape[0];
+  const int64_t n_col_out = shape[1];
+  const int64_t n_col_in = diagonals.size(1);
+  const int64_t n_diag = diagonals.size(0);
+
+  auto offsets_cont = offsets.contiguous();
+  auto diagonals_cont = diagonals.contiguous();
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  at::cuda::ThrustAllocator alloc;
+  auto policy = thrust::cuda::par(alloc).on(stream);
+  {
```

Comment:
Can you use aten operators instead of thrust operators for these or are we missing coverage?